### PR TITLE
feat: Implement ST_Crosses and ST_Overlaps predicates

### LINKED ISSUE #199: feat: Implement ST_Crosses and ST_Overlaps predicates
Here's my original commit that added the `geos` (library) implementation of the other predicates. Implementing ST_Crosses and ST_Overlaps should be as simple as pattern matching (just a library call), whatever spots you see other predicates (e.g ST_Intersects, ST_Within, etc)

https://github.com/apache/sedona-db/commit/42cca94cecc2ec85b234a2bc943015f10363fb03

For whoever picks this up, please let us know if you run into differences in behavior between the Rust implementations and PostGIS.

Part of https://github.com/apache/sedona-db/issues/174

### PR DESCRIPTION
**Pull request by Abeeujah**
closes #199

### CODE REVIEW

File: c/sedona-geos/src/binary_predicates.rs [hunk 1]
```diff
@@ -377,4 +385,81 @@ mod tests {
...
+        );
+        let arg2 = create_array(
+            &[
+                Some("POLYGON ((1 1, 1 3, 3 3, 3 1, 1 1))"),
+                Some("LINESTRING (1 0, 3 0)"),
+                Some("POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))"),
+            ],
+            &WKB_GEOMETRY,
+        );
+        let expected: ArrayRef = arrow_array!(Boolean, [Some(true), Some(true), None]);
```
Comment (paleolimbot) [inline c/sedona-geos/src/binary_predicates.rs]: I think this test would be more useful if one of the results were false

File: c/sedona-geos/src/geos.rs [hunk 2]
```diff
@@ -96,3 +96,21 @@ impl BinaryPredicate for Touches {
         lhs.touches(rhs)
     }
 }
+
+/// Check if the geometries crosses
```
Comment (paleolimbot) [inline c/sedona-geos/src/geos.rs]: nit:

```suggestion
/// Check if the geometries cross
```

File: c/sedona-geos/src/geos.rs [hunk 3]
```diff
@@ -96,3 +96,21 @@ impl BinaryPredicate for Touches {
...
+#[derive(Debug, Default)]
+pub struct Crosses {}
+impl BinaryPredicate for Crosses {
+    fn evaluate(lhs: &Geometry, rhs: &Geometry) -> GResult<bool> {
+        lhs.crosses(rhs)
+    }
+}
+
+/// Check if the geometries overlaps
```
Comment (paleolimbot) [inline c/sedona-geos/src/geos.rs]: nit: 

```suggestion
/// Check if the geometries overlap
```

**Comment by jiayuasu** [comment]
Great work!

@paleolimbot 
@Kontinuation 

how do these functions work with our current spatial join implementation?

**Comment by Kontinuation** [comment]
> how do these functions work with our current spatial join implementation?

Spatial join has [its own spatial predicate evaluators](https://github.com/apache/sedona-db/tree/main/rust/sedona-spatial-join/src/refine) for statistics awareness and better performance; UDF implementations of these predicates have no impact on spatial join.

**Comment by Abeeujah** [comment]
Hi @paleolimbot My bad, I was following the reference PR mentioned in the issue description, it didn't seem to include integrated tests on the python side, but I've currently implemented that, I'd love you to give it a look again,

For the nitpick suggestions/comments in the rust side, I've resolved them again as well, looking forward to getting a review at your earliest.

@jiayuasu and @Kontinuation Thank y'all so much for the kind words!

**Comment by petern48** [comment]
> My bad, I was following the reference PR mentioned in the issue description, it didn't seem to include integrated tests on the Python side, but I've currently implemented that, I'd love you to give it a look again

Yeah, this is my fault for not mentioning it in the issue, not yours. We added the integration test suite completely after all of the other predicates. Great job on adding them.

File: python/sedonadb/tests/functions/test_predicates.py [hunk 4]
```diff
@@ -357,3 +357,86 @@ def test_st_within_skipped(eng, geom1, geom2, expected):
...
+        ("POINT (0 0)", None, None),
+        (None, "POINT (0 0)", None),
+        ("POINT (0 0)", "LINESTRING (0 0, 1 1)", False),
+        ("LINESTRING (0 0, 2 2)", "POLYGON ((1 1, 3 1, 3 3, 1 3, 1 1))", False),
+        ("MULTIPOINT ((0 0), (1 1))", "MULTIPOINT ((1 1), (2 2))", True),
+        ("MULTIPOINT ((0 0), (1 1))", "MULTIPOINT ((0 0), (1 1))", False),
+        ("POINT (0 0)", "POINT (0 0)", False),
+        ("LINESTRING (0 0, 2 2)", "LINESTRING (1 1, 3 3)", True),
+        ("LINESTRING (0 0, 1 1)", "LINESTRING (0 1, 1 0)", False),
+        ("LINESTRING (0 0, 1 1)", "LINESTRING (1 1, 2 2)", False),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:415]: I like that we have some good edge cases here like identical geometries (`"POINT (0 0)", "POINT (0 0)"`) and "touching geometries" (i.e. `"LINESTRING (0 0, 1 1)", "LINESTRING (1 1, 2 2)"`)

File: python/sedonadb/tests/functions/test_predicates.py [hunk 5]
```diff
@@ -357,3 +357,86 @@ def test_st_within_skipped(eng, geom1, geom2, expected):
...
     )
+
+
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geom1", "geom2", "expected"),
+    [
+        (None, None, None),
+        ("POINT (0 0)", None, None),
+        (None, "POINT (0 0)", None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:368]:
```suggestion
        (None, "POINT (0 0)", None),
        ("POINT (0 0)", "POINT EMPTY", False),
```

How about we add one more case (empty geometries) while we're here. I checked out the branch and tested these locally, and saw that they already pass.

File: python/sedonadb/tests/functions/test_predicates.py [hunk 6]
```diff
@@ -357,3 +357,86 @@ def test_st_within_skipped(eng, geom1, geom2, expected):
...
+    )
+
+
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geom1", "geom2", "expected"),
+    [
+        (None, None, None),
+        ("POINT (0 0)", None, None),
+        (None, "POINT (0 0)", None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:406]:
```suggestion
        (None, "POINT (0 0)", None),
        ("POINT (0 0)", "POINT EMPTY", False),
```

**Comment by Abeeujah** [comment]
@petern48 it's all cool, your issue description and the PR linked was super helpful, gave me enough context to raise a PR, Thank you for that.
